auth/github/client.go

```diff
@@ -0,0 +1,167 @@
+/*
+Copyright 2024 The Flux authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package github
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"net/url"
+	"strconv"
+	"time"
+
+	"github.com/bradleyfalzon/ghinstallation/v2"
+	corev1 "k8s.io/api/core/v1"
+)
+
+const (
+	AppIDKey             = "githubAppID"
+	AppInstallationIDKey = "githubAppInstallationID"
+	AppPrivateKey        = "githubAppPrivateKey"
+	ApiURLKey            = "githubApiURL"
+)
+
+// Client is an authentication provider for GitHub Apps.
+type Client struct {
+	appID          int
+	installationID int
+	privateKey     []byte
+	apiURL         string
+	proxyURL       *url.URL
+	ghTransport    *ghinstallation.Transport
+}
+
+// OptFunc enables specifying options for the provider.
+type OptFunc func(*Client) error
+
+// New returns a new authentication provider for GitHub Apps.
+func New(opts ...OptFunc) (*Client, error) {
+	var err error
+
+	p := &Client{}
+	for _, opt := range opts {
+		err = opt(p)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	transport := http.DefaultTransport.(*http.Transport).Clone()
+	if p.proxyURL != nil {
+		transport.Proxy = http.ProxyURL(p.proxyURL)
```

Comment:
Validated that the requests go through proxy with proxy configuration as mentioned in https://github.com/fluxcd/flux2/issues/5062, updated the code to use this instead of proxy URL for github-app authentication. We will have to update the proxy-secret-ref [code](https://github.com/fluxcd/source-controller/blob/main/internal/controller/gitrepository_controller.go#L614) in source-controller to read the NO_PROXY data from secret and pass it to client after which it can be passed to the proxy configuration.